modules/host/negotiate.go

```diff
@@ -15,280 +15,356 @@ import (
 	h.save()
 	h.mu.Unlock(lockID)
 
-	// Send an ack to the renter that all is well.
-	err = encoding.WriteObject(conn, true)
+	return nil
+}
+
+// rpcRevise is an RPC that allows a renter to revise a file contract. It will
+// read new revisions in a loop until the renter sends a termination signal.
+func (h *Host) rpcRevise(conn net.Conn) error {
+	// read ID of contract to be revised
+	var fcid types.FileContractID
+	if err := encoding.ReadObject(conn, &fcid, crypto.HashSize); err != nil {
+		return err
+	}
+	lockID := h.mu.RLock()
+	obligation, exists := h.obligationsByID[fcid]
+	h.mu.RUnlock(lockID)
+	if !exists {
+		return errors.New("no record of that contract")
+	}
+
+	// need to protect against two simultaneous revisions to the same
+	// contract; this can cause inconsistency and data loss, making storage
+	// proofs impossible
+	obligation.mu.Lock()
+	defer obligation.mu.Unlock()
+
+	// open the file (or create it if it does not exist)
+	file, err := os.OpenFile(obligation.Path, os.O_WRONLY|os.O_APPEND, 0660)
 	if err != nil {
-		return
+		return err
+	}
+
+	// rebuild current Merkle tree
+	tree := crypto.NewTree()
+	buf := make([]byte, crypto.SegmentSize)
+	for {
+		_, err := io.ReadFull(file, buf)
+		if err == io.EOF {
+			break
+		} else if err != nil && err != io.ErrUnexpectedEOF {
+			return err
+		}
+		tree.Push(buf)
+	}
+
+	// accept new revisions in a loop
+	for {
+		// read proposed revision
+		var rev types.FileContractRevision
+		if err := encoding.ReadObject(conn, &rev, types.BlockSizeLimit); err != nil {
+			return err
+		}
+		// an empty revision indicates completion
+		if rev.ParentID == (types.FileContractID{}) {
+			break
+		}
+
+		// check revision against original file contract
+		lockID = h.mu.RLock()
+		err := h.considerRevision(rev, obligation)
+		height := h.blockHeight
+		h.mu.RUnlock(lockID)
+		if err != nil {
+			encoding.WriteObject(conn, err.Error())
+			continue // don't terminate loop; subsequent revisions may be okay
+		}
+
+		// indicate acceptance
+		if err := encoding.WriteObject(conn, modules.AcceptResponse); err != nil {
+			return err
+		}
+
+		// read piece
+		piece := make([]byte, rev.NewFileSize-obligation.FileContract.FileSize)
+		_, err = io.ReadFull(conn, piece)
+		if err != nil {
+			return err
+		}
+
+		// verify Merkle root
+		tree.Push(piece)
+		if tree.Root() != rev.NewFileMerkleRoot {
+			return errors.New("revision has bad Merkle root")
+		}
+
+		// read signed transaction
+		var signedTxn types.Transaction
+		if err := encoding.ReadObject(conn, &signedTxn, types.BlockSizeLimit); err != nil {
+			return err
+		}
+
+		// check signature
+		if err := signedTxn.StandaloneValid(height); err != nil {
+			return err
+		}
+
+		// sign and return transaction
+		txnBuilder := h.wallet.RegisterTransaction(signedTxn, nil)
+		for _, sig := range signedTxn.TransactionSignatures {
+			txnBuilder.AddTransactionSignature(sig)
+		}
+		signedTxnSet, err := txnBuilder.Sign(true)
```

Comment:
the transaction builder would also need some way to figure out which key it's supposed to be signing.

Probably best if I write the signing code. The process may look daunting, but I think it's ultimately pretty simple:

1. use the CoveredFields object to indicate what parts of the transaction should get covered by the signature.
2. create the TransactionSignature object to indicate which object + public key the signature is signing (and also supply the covered fields from 1).
3. call txn.SigHash to get the object you are supposed to sign.
4. call crypto.SignHash to create the cryptographic signature
5. add the cryptographic signature to the transaction.

While it is a lot of steps, the process is pretty straigtforward I think.